### PR TITLE
Allow password definition during VM deploy

### PR DESCRIPTION
**Pull request by stephankruggg**
### Description

This PR allows users to define the password of a virtual machine when deploying the VM (`deployVirtualMachine` API). If no password is informed, current ACS behavior is maintained: a random password is generated, injected into the VM, and returned to the user.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

### How Has This Been Tested?

I applied the changes in a local lab, called the `deployVirtualMachine` API defining a password, and successfully accessed the VM with the password I have defined. Additionally, I called the `deployVirtualMachine` API without defining a password, and successfully accessed the VM with the random password generated.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
please guard debug logging statements if they need parameter evaluation. code looks good otherwise I think, but I will not approve logging like this. heavy loaded systems ar benefitted by not having the extra string operations when they don´t want to.

**Review by JoaoJandre — APPROVED** [review]
LGTM

File: api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java [hunk 1]
```diff
@@ -103,6 +103,9 @@ public class DeployVMCmd extends BaseAsyncCreateCustomIdCmd implements SecurityG
     @Parameter(name = ApiConstants.DISPLAY_NAME, type = CommandType.STRING, description = "an optional user generated name for the virtual machine")
     private String displayName;
 
+    @Parameter(name=ApiConstants.PASSWORD, type=CommandType.STRING, description="The password of the virtual machine. If null, a random password will be generated for the VM.")
```
Comment (rohityadavcloud) [inline api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java:106]: Can you add a `since` param in the @Parameter

**Review by RodrigoDLopez — CHANGES_REQUESTED** [review]
Hey @stephankruggg   
During tests with this proposal, I noticed an NPE when restarting a VM with the force option checked. Can you take a look at this?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -5398,11 +5403,19 @@ public Pair<UserVmVO, Map<VirtualMachineProfile.Param, Object>> startVirtualMach
             String password = "saved_password";
             if (template.isEnablePassword()) {
                 if (vm.getDetail("password") != null) {
+                    s_logger.debug(String.format("Decrypting VM [%s] current password.", vm));
                     password = DBEncryptionUtil.decrypt(vm.getDetail("password"));
+                } else if (additionalParams.containsKey(VirtualMachineProfile.Param.VmPassword)) {
```
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:5408]: good catch @RodrigoDLopez , but I think the null check must go upstream a bit as there is another use of the `additionalParams` below.
I think it pays (extra) to factor this bit of code out into a new method @stephankruggg

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 3]
```diff
@@ -1158,9 +1159,8 @@ private UserVm rebootVirtualMachine(long userId, long vmId, boolean enterSetup,
     private UserVm forceRebootVirtualMachine(long vmId, long hostId, boolean enterSetup) {
         try {
             if (stopVirtualMachine(vmId, false) != null) {
-                Map<VirtualMachineProfile.Param,Object> params = null;
+                Map<VirtualMachineProfile.Param,Object> params = new HashMap();
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:1162]:
```suggestion
                Map<VirtualMachineProfile.Param,Object> params = new HashMap<>();
```

**Comment by DaanHoogland** [comment]
@stephankruggg please look at the code smell, a diamond operator missing,

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 4]
```diff
@@ -134,6 +134,7 @@
 import org.apache.commons.lang3.builder.ToStringBuilder;
 import org.apache.commons.lang3.builder.ToStringStyle;
 import org.apache.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:137]: This is very convenient, but should we allow jetbrains specific annotations? cc @rohityadavcloud @GutoVeronezi @weizhouapache @wido

Comment (weizhouapache) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:137]: I am ok with it

Comment (GutoVeronezi) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:137]: I am OK with it as well; the project [JetBrains/java-annotations](https://github.com/JetBrains/java-annotations) is highly used, receives upgrades periodically, and the annotations are well documented.

**Comment by DaanHoogland** [comment]
@RodrigoDLopez , are your findings dealt with ok?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 3 above)
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:1162]: please add the diamond operator

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by RodrigoDLopez — APPROVED** [review]
LGTM, after manually tested and code review.

**Comment by GaOrtiga** [comment]
Hello, I spoke with @stephankruggg in person, and, in order to keep the flow going, I will help him with any conflicts/change requests that are pending on this PR.

File: api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java [hunk 5]
```diff
@@ -105,6 +105,10 @@ public class DeployVMCmd extends BaseAsyncCreateCustomIdCmd implements SecurityG
     @Parameter(name = ApiConstants.DISPLAY_NAME, type = CommandType.STRING, description = "an optional user generated name for the virtual machine")
     private String displayName;
 
+    @Parameter(name=ApiConstants.PASSWORD, type=CommandType.STRING, description="The password of the virtual machine. If null, a random password will be generated for the VM.",
+            since="4.18.0.0")
```
Comment (gpordeus) [inline api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java:109]:
```suggestion
            since="4.19.0.0")
```

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@stephankruggg @GaOrtiga can you look at rebasing this, the unit tests fail:
```
16:40:33 [ERROR] Errors: 
16:40:33 [ERROR]   AutoScaleManagerImplTest.unnecessary Mockito stubbings ? UnnecessaryStubbing 
```

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by GaOrtiga** [comment]
> @stephankruggg @GaOrtiga can you look at rebasing this, the unit tests fail:
> 
> ```
> 16:40:33 [ERROR] Errors: 
> 16:40:33 [ERROR]   AutoScaleManagerImplTest.unnecessary Mockito stubbings ? UnnecessaryStubbing 
> ```

Yes, I will look into it

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan LLtest keepEnv

**Review by DaanHoogland — APPROVED** [review]
note this is only on the API (not the UI)
tested ok.
clgtm

**Pull request merged**